File: M08_Architecture/C02_Component/S02_LiveData/src/main/java/net/bi4vmr/study/base/MyViewModel.java
package net.bi4vmr.study.base;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

/**
 * Name        : MyViewModel
 * <p>
 * Author      : BI4VMR
 * <p>
 * Email       : bi4vmr@outlook.com
 * <p>
 * Date        : 2023-05-11 14:51
 * <p>
 * Description : ViewModel：基本功能测试。
 */
public class MyViewModel extends ViewModel {

    private static final String TAG = "TestApp-" + MyViewModel.class.getSimpleName();

    // 基本类型数值
    private int num = 0;

    // 可变LiveData，其中的数值可以被修改。
    private final MutableLiveData<Integer> numberData = new MutableLiveData<>();
    // 不可变LiveData，仅可被外部观察。
    public final LiveData<Integer> roNumberData = numberData;

    // 数值增加
    public void plus() {
        Log.i(TAG, "Plus.");
        // 改变数值
        num += 10;
        // 通知观察者数值发生变化
        numberData.setValue(num);
    }

    // 数值减少
    public void minus() {
        Log.i(TAG, "Minus.");
        // 改变数值
        num -= 10;
        // 通知观察者数值发生变化
        numberData.setValue(num);
    }
}

File: M03_UI/C10_Compose/S04_State/src/main/kotlin/net/bi4vmr/study/statesave/MyViewModel.kt
package net.bi4vmr.study.statesave

import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

/**
 * ViewModel：计数器示例。
 *
 * @author bi4vmr@outlook.com
 * @since 1.0.0
 */
class MyViewModel : ViewModel() {

    // 可变 StateFlow ，仅内部使用。
    private val _count: MutableStateFlow<Int> = MutableStateFlow(0)

    // 不可变 StateFlow ，与可变实例保持同步，供外部观察。
    val count: StateFlow<Int> = _count


    // 业务接口：数值增加。
    fun add() {
        _count.value++
    }

    // 业务接口：数值减少。
    fun reduce() {
        _count.value--
    }
}
